File: bc-w3/bcw3/Basic/Point/Point.cpp
#include <iostream>
#include <cmath>
#include "Point.h"

Point::Point(double x, double y) : x(x), y(y) {}

Point::~Point() {}

double Point::getX() const {
    return this->x;
}

double Point::getY() const {
    return this->y;
}

void Point::setX(double value) {
    this->x = value;
}

void Point::setY(double value) {
    this->y = value;
}

double Point::distance(const Point& other) const {
    return hypot(other.x-this->x, this->y-other.y);
}

bool Point::operator==(const Point& other) const {
    return other.x == this->x && other.y == this->y; 
}

bool Point::operator!=(const Point& other) const {
    return !(*this == other);
}

std::ostream& operator<<(std::ostream& out, const Point& point) {
    return out << "(" << point.getX() << ", " << point.getY() << ")";
}

File: bc-w3/java/Basic/Point/Point.java
import java.text.DecimalFormat;

public class Point implements Cloneable {
    private double x;
    private double y;
    
    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
    
    public Point() {
        this(0, 0);
    }
    
    public double getX() {
        return x;
    }
    
    public double getY() {
        return y;
    }
    
    public void setX(double value) {
        this.x = value;
    }
    
    public void setY(double value) {
        this.y = value;
    }
    
    public double distance(Point other) {
        return Math.hypot(other.x-this.x, this.y-other.y);
    }
    
    public boolean equals(Point other) {
        return other.x == this.x && other.y == this.y;
    }
    
    public String toString() {
        DecimalFormat fmt = new DecimalFormat("######.######");
        
        return "(" + fmt.format(this.x) + ", " + fmt.format(this.y) + ")";
    }
    
    public Point clone() throws CloneNotSupportedException {
        return (Point) super.clone();
    }
}
